sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpanderUtil.java

```diff
@@ -345,27 +326,157 @@ private void expand(
+        /** The collapser is present to allow each stack frame to handle forks. */
+        private final ApexPathCollapser apexPathCollapser;
+
+        PendingForkStack(ApexPathCollapser apexPathCollapser) {
+            stack = new Stack<>();
+            this.apexPathCollapser = apexPathCollapser;
+        }
+
+        /**
+         * Use the provide Exception to create a new {@link PendingForkStackFrame} and add it to the
+         * top of the stack.
+         *
+         * @param ex An exception thrown when a method is determined to fork.
+         */
+        void addStackFrame(MethodPathForkedException ex) {
+            stack.push(new PendingForkStackFrame(ex, apexPathCollapser));
+        }
+
+        /**
+         * @return An Optional containing either the next {@link ApexPathExpander} to be processed,
+         *     or nothing.
+         */
+        Optional<ApexPathExpander> getNextExpander() {
+            // Base Case: The stack is empty, so there are no more frames to inspect.
+            // Return an empty optional.
+            if (stack.empty()) {
+                return Optional.empty();
+            }
+            PendingForkStackFrame frame = stack.peek();
+            Optional<ApexPathExpander> nextExpanderOptional = frame.getNextExpander();
+            if (nextExpanderOptional.isEmpty()) {
+                // Recursive Case: The current frame is empty. Pop it off, and call recursively
+                // to move to the next frame.
+                stack.pop();
+                return getNextExpander();
+            } else {
+                // Base Case: The current frame returned an Expander. Return it.
+                return nextExpanderOptional;
+            }
+        }
+    }
+
+    /**
+     * Each frame in the {@link PendingForkStack} represents a fork event currently being processed.
+     */
+    private static class PendingForkStackFrame {
+        /** The exception that triggered the fork. */
+        private final MethodPathForkedException ex;
+        /** The collapser used to process forks. */
+        private final ApexPathCollapser apexPathCollapser;
+        /**
+         * The index of the next {@link ApexPath} that should be turned into an {@link
+         * ApexPathExpander}.
+         */
+        private int idx;
+
+        PendingForkStackFrame(MethodPathForkedException ex, ApexPathCollapser apexPathCollapser) {
+            this.ex = ex;
+            this.apexPathCollapser = apexPathCollapser;
+            this.idx = 0;
+        }
+
+        /**
+         * @return {@code true} if this frame is out of paths to turn into expanders; else {@link
+         *     false}.
+         */
+        private boolean done() {
+            return idx >= ex.getPaths().size();
+        }
+
+        /**
+         * @return An Optional containing either the next {@link ApexPathExpander} or nothing.
+         */
+        Optional<ApexPathExpander> getNextExpander() {
```

Comment:
The contents of this method are derived from the old `getForkedExpanders()` and `catch`-block.